### ISSUE TITLE
Is "hexo deploy" delete old files on remote host first? then push? why do this?

### ISSUE THREAD

**Issue by xu-song** [hexojs/hexo#3002]
## Environment Info

Node version(`node -v`):
v8.9.4
Your site `_config.yml` (Optional):
```
deploy:
  type: git
  repo: git@github.com:xu-song/xu-song.github.io.git
  branch: master
```
Your theme `_config.yml` (Optional): 

Hexo and Plugin version(`npm ls --depth 0`): 

```
├── hexo@3.5.0
├── hexo-deployer-git@0.3.1
├── hexo-generator-archive@0.1.5
├── hexo-generator-category@0.1.3
├── hexo-generator-index@0.2.1
├── hexo-generator-search@2.1.1
├── hexo-generator-searchdb@1.0.8
├── hexo-generator-tag@0.2.0
├── hexo-renderer-ejs@0.3.1
├── hexo-renderer-marked@0.3.2
├── hexo-renderer-stylus@0.3.3
└── hexo-server@0.2.2
```

## For question

`hexo deploy` will delete old files on remote host. right?
why do this?  more convenient to push?

[doc reference](https://hexo.io/docs/deployment.html)

**Comment by NoahDragon** [comment]
You are right. The hexo git deployer will do a force push.

**Comment by xu-song** [comment]
`delete old files on remote host first`. 
The above statement is not  accurate technically. It may cause misunderstanding.

According to the [**deploy.js source code**](https://github.com/hexojs/hexo-deployer-git/blob/master/lib/deployer.js),
`hexo d` equals to the following git command, in most situation.
```
$ git add -A
$ git commit -m "Site updated: 2018-01-30 *:*:*" 
$ git push -u origin HEAD:master --force
```

It just deletes the conflicting commits from remote, not all old files.  Right?

**Comment by NoahDragon** [comment]
Actually, the `delete old files on remote host first` is literally correct. Even the commit history is also lost.

**Comment by xu-song** [comment]
I think `overwrite` is a more [official statement](https://git-scm.com/docs/git-push) instead of `delete all`. I am not sure whether or not `git force push` will delete old files first.

**Comment by wololock** [comment]
Commits history gets lost if you run `hexo deploy` from a folder that does not contain current `.deploy_git` folder. It happens e.g. if you build and deploy your blog using continuous integration server (like Travis CI for instance) that always works on a fresh cloned copy.

I have faced this problem when I started using Travis CI to run `hexo deploy`. A solution that works for me is pretty simple - I clone the branch with generated HTML to `.deploy_git` before I execute `hexo deploy` - https://github.com/wololock/wololock.github.io/blob/develop/.travis.yml#L30
